src/refract/JSONSchemaVisitor.cc

```diff
@@ -7,646 +7,664 @@
-    {
-        std::set<std::string> required;
+                const StringElement* str = TypeQueryVisitor::as<const StringElement>(mr->get().key());
 
-        for (const auto& member: members) {
-            if (!member) {
-                continue;
+                if (str) {
+                    required.insert(str->get());
+                }
             }
 
-            TypeQueryVisitor type;
-            Visit(type, *member);
-
-
-            switch (type.get()) {
-                case TypeQueryVisitor::Member: {
-                    MemberElement* mr = static_cast<MemberElement*>(member);
+            if (IsVariableProperty(*mr->get().key())) {
+                varProps.push_back(mr);
+            } else {
+                JSONSchemaVisitor renderer(pDefs, fixed);
+                Visit(renderer, member);
+                ObjectElement* o1 = TypeQueryVisitor::as<ObjectElement>(renderer.get());
+
+                if (!o1->get().empty()) {
+                    auto front = o1->get().begin()[0]->clone();
+                    if (TypeQueryVisitor::as<MemberElement>(front.get())) {
+                        if (o.empty())
+                            o.set();
+                        o.get().push_back(std::move(front));
+                    }
+                }
+            }
+        } break;
 
-                    if (IsTypeAttribute(*member, "required") || IsTypeAttribute(*member, "fixed")
-                        || ((fixed || fixedType) && !IsTypeAttribute(*member, "optional"))) {
+        case TypeQueryVisitor::Select: {
+            const SelectElement* sel = static_cast<const SelectElement*>(&member);
 
-                        StringElement* str = TypeQueryVisitor::as<StringElement>(mr->value.first);
+            // FIXME: there is no valid solution for multiple "SelectElement" in one object.
 
-                        if (str) {
-                            required.insert(str->value);
-                        }
-                    }
+            for (auto const& select : sel->get()) {
+                JSONSchemaVisitor v(pDefs);
+                VisitBy(*select, v);
+                oneOfMembers.push_back(v.getOwnership());
+            }
+        } break;
 
-                    if (IsVariableProperty(*mr->value.first)) {
-                        varProps.push_back(mr);
-                    } else {
-                        JSONSchemaVisitor renderer(pDefs, fixed);
-                        Visit(renderer, *member);
-                        ObjectElement* o1 = TypeQueryVisitor::as<ObjectElement>(renderer.get());
+        default:
+            throw LogicError("Invalid member type of object in MSON definition");
+    }
+}
 
-                        if (!o1->value.empty()) {
-                            MemberElement* m1 = TypeQueryVisitor::as<MemberElement>(o1->value[0]->clone());
+std::string refract::renderJsonSchema(const IElement& e)
+{
+    auto pDefs = make_element<ObjectElement>();
+    JSONSchemaVisitor gen_schema(*pDefs);
```

Comment:
While naming in new parts is possible to discuss, in old code base I'm absolutely against.

Reply:
`gen_schema` to `generate`?

Reply:
Still it is local variable. We have naming conventions for it

Reply:
`generate` is ok, sorry